costs/es_request_parser.go

```diff
@@ -0,0 +1,128 @@
+	key, ok := in["key_as_string"]
+	if !ok {
+		key, ok = in["key"]
+	}
+	if skey := fmt.Sprintf("%v", key); ok {
+		if aggrMapType, ok := aggrType.(map[string]interface{}); ok && aggrName[0] == '*' {
+			aggrType := aggrMapType["buckets"]
+			flattenSubAggregation(&aggrType)
+		}
+		in[skey] = aggrType
+		delete(in, "key")
+		delete(in, "key_as_string")
+	}
+	delete(in, aggrName)
+}
+
+// browseMapRecursivly browses a map and calls browseMapRecursivly or
+// browseTabRecursivly, and stripAggregationArchitecture for each fields.
+func browseMapRecursivly(in map[string]interface{}) {
+	for aggrName, field := range in {
+		if aggrName[0] != '&' && aggrName[0] != '*' {
+			continue
+		}
+		field := field.(map[string]interface{})
+		metadata := strings.Split(aggrName[1:], ".")
+		aggrTypeName := strings.Replace(metadata[0], ",", ".", -1)
+		aggrType := field[aggrTypeName]
+		if aggrName[0] == '&' {
+			browseTabRecursivly(&aggrType)
+		} else if aggrName[0] == '*' {
+			browseMapRecursivly(field)
+		}
+		stripAggregationArchitecture(aggrName, aggrType, in)
+	}
+}
+
+// browseTabRecursivly browses a map and calls browseMapRecursivly.
+// In the end it calls flattenSubAggregation to flat the array.
+func browseTabRecursivly(in *interface{}) {
+	tab := (*in).([]interface{})
+	for _, field := range tab {
+		field := field.(map[string]interface{})
+		delete(field, "doc_count")
+		browseMapRecursivly(field)
+	}
+	flattenSubAggregation(in)
+}
+
+// prepareRecursiveParsing prepares the first run of the recursive and
+// calls browseMapRecursivly or browseTabRecursivly.
+func prepareRecursiveParsing(result map[string]interface{}, aggrName string, aggr *json.RawMessage) error {
+	var root map[string]interface{}
+	err := json.Unmarshal(*aggr, &root)
+	if err != nil {
+		return err
+	}
+	metadata := strings.Split(aggrName[1:], ".")
+	name := metadata[len(metadata)-1]
+	aggrTypeName := strings.Replace(metadata[0], ",", ".", -1)
+	aggrType := root[aggrTypeName]
+	if aggrName[0] == '&' {
+		browseTabRecursivly(&aggrType)
+		result[name] = aggrType
+	} else if aggrName[0] == '*' {
+		browseMapRecursivly(root)
+		bucketAggr := aggrType.(map[string]interface{})["buckets"]
+		flattenSubAggregation(&bucketAggr)
+		result[name] = bucketAggr
+	} else {
+		result[aggrName] = root
+	}
+	return nil
+}
+
+// GetJSONSimplifiedElasticSearchResult is used to parse an *elastic.SearchResult
+// to a human readable map[string]interface{} that will be able to be
+// transformed to JSON and send to the UI.
+// It will only parse aggregations with the name &buckets.{{name}} or *value.{{name}}
+// i.e. &buckets.regions and *value.cost.
+func GetJSONSimplifiedElasticSearchResult(esResult *elastic.SearchResult) (map[string]interface{}, error) {
+	res := make(map[string]interface{})
+	for name, aggregation := range esResult.Aggregations {
+		if err := prepareRecursiveParsing(res, name, aggregation); err != nil {
```

Comment:
Why call the function `prepareRecursiveParsing` if it does the whole job ? When I see a `prepare*` function I expect that it should be followed with the logic which actually performs the job. Also this function seems to act on its arguments, which is usually Bad™. You can either change `prepareRecursiveParsing` to return its result, or define a type underlaid by `map[string]interface{}` and define methods over this type. At the very least, make sure it is _obvious_ that your function modifies its arguments when it does.